client/src/main/java/io/pravega/client/tables/impl/TableSegment.java

```diff
@@ -0,0 +1,172 @@
+     * @param key  The Key to remove. If {@link TableKey#getVersion()} is null, this will perform an Unconditional Remove,
+     *             otherwise it will perform a Conditional Remove based on the information provided.
+     *             See {@link TableSegment} doc for more details on Types of Updates.
+     * @return A CompletableFuture that, when completed, will indicate the Key has been removed. Notable exceptions:
+     * <ul>
+     * <li>{@link ConditionalTableUpdateException} If this is a Conditional Removal and the condition was not satisfied.
+     * </ul>
+     */
+    CompletableFuture<Void> remove(TableKey<KeyT> key);
+
+    /**
+     * Removes one or more keys from this Table Segment.
+     *
+     * @param keys A Collection of keys to remove. If for at least one such key, {@link TableKey#getVersion()} returns
+     *             a non-null value, this will perform an atomic Conditional Remove where all the keys either get removed
+     *             or none will; otherwise an Unconditional Remove will be performed. See {@link TableSegment} doc for more
+     *             details on Types of Updates.
+     * @return A CompletableFuture that, when completed, will indicate that the keys have been removed. Notable exceptions:
+     * <ul>
+     * <li>{@link ConditionalTableUpdateException} If this is a Conditional Removal and the condition was not satisfied.
+     * This exception will contain the Keys that failed the validation.
+     * </ul>
+     */
+    CompletableFuture<Void> remove(Collection<TableKey<KeyT>> keys);
+
+    /**
+     * Gets the latest value for the given Key.
+     *
+     * @param key The Key to get the value for.
+     * @return A CompletableFuture that, when completed, will contain the requested result. If no such Key exists, this
+     * will be completed with a null value.
+     */
+    CompletableFuture<TableEntry<KeyT, ValueT>> get(KeyT key);
+
+    /**
+     * Gets the latest values for the given Keys.
+     *
+     * @param keys A Collection of Keys to get values for.
+     * @return A CompletableFuture that, when completed, will contain a map of {@link KeyT} to {@link TableEntry} for those
+     * keys that have a value in the index. All other keys will not be included.
+     */
+    CompletableFuture<Map<KeyT, TableEntry<KeyT, ValueT>>> get(Collection<KeyT> keys);
+
+    /**
+     * Creates and registers a {@link KeyUpdateListener} for all updates to this {@link TableSegment}, subject to the given
+     * {@link KeyUpdateFilter}.
+     *
+     * @param filter   A {@link KeyUpdateFilter} that will specify which Keys should the {@link KeyUpdateListener} listen to.
+     * @param executor An {@link Executor} that will be used to invoke all callbacks on.
+     * @return A CompletableFuture that, when completed, will contain an {@link KeyUpdateListener}.
+     */
+    CompletableFuture<KeyUpdateListener<KeyT, ValueT>> createListener(KeyUpdateFilter<KeyT> filter, Executor executor);
```

Comment:
I think the guarantees of such listener are not clear. I read the javadoc for `KeyUpdateListener` and it refers to correctness from a concurrency perspective. I'm wondering if the updates are guaranteed to be delivered even in the presence of crashes of Pravega components.

Reply:
This will be addressed in #2882 . It's too early to think through all the details of this feature.